### PR TITLE
Cache ItemMeta instead of creating a new ItemMeta for each potion check

### PR DESCRIPTION
**Pull request by MrPowerGamerBR**
The `isSimilarPotion` is very expensive because the code attempts to get the `otherPotion` (in this case, the current item) `ItemMeta` on every comparison check, and initializing the `ItemMeta` is *very* expensive.

To avoid this, we only create the `ItemMeta` once and then reuse it in the `isSimilarPotion` check.

While on Spigot the ItemStack code does seem to cache the ItemMeta, on Paper it doesn't due to the ItemStack being delegated to the server ItemStack, the ItemMeta seems to be always be created on every `getItemMeta()` call.

image

### CODE REVIEW

File: src/main/java/com/gmail/nossr50/datatypes/skills/alchemy/AlchemyPotion.java [hunk 1]
```diff
@@ -18,13 +19,15 @@
...
     private final @NotNull ItemStack potionItemStack;
+    private final @NotNull ItemMeta potionItemMeta;
     private final @NotNull Map<ItemStack, String> alchemyPotionChildren;
 
     public AlchemyPotion(@NotNull String potionConfigName, @NotNull ItemStack potionItemStack,
                          @NotNull Map<ItemStack, String> alchemyPotionChildren) {
         this.potionConfigName = requireNonNull(potionConfigName, "potionConfigName cannot be null");
         this.potionItemStack = requireNonNull(potionItemStack, "potionItemStack cannot be null");
         this.alchemyPotionChildren = requireNonNull(alchemyPotionChildren, "alchemyPotionChildren cannot be null");
+        this.potionItemMeta = potionItemStack.getItemMeta(); // The potion item meta should never be null because it is a potion
```
Comment (nossr50) [inline src/main/java/com/gmail/nossr50/datatypes/skills/alchemy/AlchemyPotion.java:30]: add `requireNonNull`

File: src/main/java/com/gmail/nossr50/config/skills/alchemy/PotionConfig.java [hunk 2]
```diff
@@ -334,9 +335,14 @@ public AlchemyPotion getPotion(String name) {
      * @return AlchemyPotion that corresponds to the given ItemStack.
      */
     public AlchemyPotion getPotion(ItemStack item) {
+        // Fast return if the item does not have any item meta to avoid initializing an unnecessary ItemMeta instance
```
Comment (nossr50) [inline src/main/java/com/gmail/nossr50/config/skills/alchemy/PotionConfig.java:338]: Smart, I like this change.

File: src/main/java/com/gmail/nossr50/datatypes/skills/alchemy/AlchemyPotion.java [hunk 3]
```diff
@@ -49,21 +52,26 @@ public AlchemyPotion(@NotNull String potionConfigName, @NotNull ItemStack potion
...
+    public boolean isSimilarPotion(@NotNull ItemStack otherPotion, @Nullable ItemMeta otherMeta) {
         requireNonNull(otherPotion, "otherPotion cannot be null");
+
         if (otherPotion.getType() != potionItemStack.getType()) {
             return false;
         }
 
         // no potion meta, no match
-        if (!otherPotion.hasItemMeta()) {
+        if (otherMeta == null) {
```
Comment (nossr50) [inline src/main/java/com/gmail/nossr50/datatypes/skills/alchemy/AlchemyPotion.java:66]: Don't understand why this change is necessary, isn't it equivalent to the prior version?

**Review by nossr50 — CHANGES_REQUESTED** [review]
Looks good, small changes requested.

File: src/main/java/com/gmail/nossr50/datatypes/skills/alchemy/AlchemyPotion.java (hunk 3 above)
Comment (MrPowerGamerBR) [reply src/main/java/com/gmail/nossr50/datatypes/skills/alchemy/AlchemyPotion.java:66]: It is equivalent, but iirc I added that check because IDEA complains that you are using a maybe null value in `hasDifferingCustomEffects`... or maybe it was my Kotlin brain saying that I should add a null check.

Comment (MrPowerGamerBR) [reply src/main/java/com/gmail/nossr50/datatypes/skills/alchemy/AlchemyPotion.java:66]: image

**Comment by nossr50** [comment]
Thank you, I appreciate it!
